File: src/main/java/com/github/qky666/selenidepom/ElementsContainerWidget.java
package com.github.qky666.selenidepom;

import com.codeborne.selenide.ElementsContainer;
import com.codeborne.selenide.SelenideElement;
import lombok.NoArgsConstructor;

import javax.annotation.CheckReturnValue;
import java.lang.reflect.Field;

/**
 * ElementsContainer with a constructor that sets 'self' field.
 */
@NoArgsConstructor
public abstract class ElementsContainerWidget extends ElementsContainer implements RequiredContainer {
    @CheckReturnValue
    public ElementsContainerWidget(SelenideElement self) {
        super();

        // Set self
        Class<?> superclass = ElementsContainerWidget.class.getSuperclass();
        Field field = null;
        try {
            field = superclass.getDeclaredField("self");
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        }
        assert field != null;
        field.setAccessible(true);
        try {
            field.set(this, self);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        assert field.isAccessible();
    }
}

File: src/main/kotlin/com/github/qky666/selenidepom/kotlin/ElementsContainerWidget.kt
@file:Suppress("unused")

package com.github.qky666.selenidepom.kotlin

import com.codeborne.selenide.ElementsContainer
import com.codeborne.selenide.SelenideElement

/**
 * [ElementsContainer] with a constructor that sets 'self' field.
 */
@Suppress("LeakingThis")
abstract class ElementsContainerWidget() : ElementsContainer(), RequiredContainer {
    constructor(self: SelenideElement) : this() {
    // Set self
        val superclass: Class<*> = ElementsContainerWidget::class.java.superclass
        val field = superclass.getDeclaredField("self")
        field.isAccessible = true
        field[this] = self
    }
}
